cmd/agent/app/jmx.go

```diff
@@ -0,0 +1,184 @@
+		Short: "Start the collection of metrics based on your current configuration and display them in the console.",
+		Long:  ``,
+		Run:   doJmxCollect,
+	}
+
+	jmxListEverythingCmd = &cobra.Command{
+		Use:   "everything",
+		Short: "List every attributes available that has a type supported by JMXFetch.",
+		Long:  ``,
+		Run:   doJmxListEverything,
+	}
+
+	jmxListMatchingCmd = &cobra.Command{
+		Use:   "matching",
+		Short: "List attributes that match at least one of your instances configuration.",
+		Long:  ``,
+		Run:   doJmxListMatching,
+	}
+
+	jmxListLimitedCmd = &cobra.Command{
+		Use:   "limited",
+		Short: "List attributes that do match one of your instances configuration but that are not being collected because it would exceed the number of metrics that can be collected.",
+		Long:  ``,
+		Run:   doJmxListLimited,
+	}
+
+	jmxListCollectedCmd = &cobra.Command{
+		Use:   "collected",
+		Short: "List attributes that will actually be collected by your current instances configuration.",
+		Long:  ``,
+		Run:   doJmxListCollected,
+	}
+
+	jmxListNotMatchingCmd = &cobra.Command{
+		Use:   "not-matching",
+		Short: "List attributes that don’t match any of your instances configuration.",
+		Long:  ``,
+		Run:   doJmxListNotCollected,
+	}
+
+	checks = []string{}
+)
+
+func init() {
+	// attach list and collect commands to jmx command
+	jmxCmd.AddCommand(jmxListCmd)
+	jmxCmd.AddCommand(jmxCollectCmd)
+
+	//attach list commands to list root
+	jmxListCmd.AddCommand(jmxListEverythingCmd, jmxListMatchingCmd, jmxListLimitedCmd, jmxListCollectedCmd, jmxListNotMatchingCmd)
+
+	jmxListCmd.PersistentFlags().StringSliceVar(&checks, "checks", []string{"jmx"}, "JMX checks (ex: jmx,tomcat)")
+	jmxCollectCmd.PersistentFlags().StringSliceVar(&checks, "checks", []string{"jmx"}, "JMX checks (ex: jmx,tomcat)")
+
+	// attach the command to the root
+	AgentCmd.AddCommand(jmxCmd)
+}
+
+func doJmxCollect(cmd *cobra.Command, args []string) {
+	runJmxCommand("collect")
+}
+
+func doJmxListEverything(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_everything")
+}
+
+func doJmxListMatching(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_matching_attributes")
+}
+
+func doJmxListLimited(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_limited_attributes")
+}
+
+func doJmxListCollected(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_collected_attributes")
+}
+
+func doJmxListNotCollected(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_not_matching_attributes")
+}
+
+func runJmxCommand(command string) {
+	err := common.SetupConfig(confFilePath)
+	if err != nil {
+		log.Fatalln("unable to set up global agent configuration: %v", err)
+	}
+
+	common.SetupAutoConfig(config.Datadog.GetString("confd_path"))
+
+	runner := jmxfetch.New()
+
+	runner.ReportOnConsole = true
+	runner.Command = command
+
+	dir, err := ioutil.TempDir(os.TempDir(), "jmxconfd")
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer os.RemoveAll(dir)
```

Comment:
deferred calls won't be executed if `os.Exit` is called, which is what `log.Fatal` does.

Alternative: make the commands return an `error`, use `cobra.Command{RunE: [...]}`. Commands return an error whenever they run into "fatal" errors like this, and cobra takes cares of printing the error message.